content/blog/unsafe-string-interning-in-go.md

```diff
@@ -0,0 +1,345 @@
+	val[cmpVal] = v
+	return v
+}
+```
+
+Usage of the package is straightforward at this point:
+
+```go
+func TestBasics(t *testing.T) {
+	foo := intern.Get("foo")
+	foo2 := intern.Get("foo")
+
+	if foo.Get() != foo2.Get() {
+		t.Error("foo values differ")
+	}
+	if foo.Get() != "foo" {
+		t.Error("foo.Get not foo")
+	}
+	if foo != foo2 {
+		t.Error("foo pointers differ")
+	}
+}
+```
+
+The global pool in the library stored the string `"foo"` on the first call to
+`intern.Get("foo")` and the same `*intern.Value` was returned and stored in both
+`foo` and `foo2`.
+
+However there is a caveat to this approach! If we keep adding more and more
+values to the pool, it will continue to grow in size and allocate more and more
+memory until our program eventually runs out of memory and crashes.
+
+In fact, this is the tradeoff made by `go4.org/intern` for the "safe" operating
+mode of the package. If you run your program with the environment variable
+`GO4_INTERN_SAFE_BUT_LEAKY=true` then no unsafe tricks will be used, and the
+intern pool will continue to grow as more items are added. For some programs,
+this may be an acceptable tradeoff, but there's much more to be discussed!
+
+## Breaking the unsafe rules
+
+In order to make use of string interning while also cooperating with Go's
+garbage collector to remove and free unused values from the pool, it is
+necessary to bend the package `unsafe` rules. This is what inspired me to author
+this blog: I've written a lot of unsafe code in the past but have never
+intentionally stepped outside the guard rails that the [`unsafe.Pointer` rules
+provide for you](https://golang.org/pkg/unsafe/#Pointer).
+
+Specifically, we have to violate rule number 3:
+
+> (3) Conversion of a Pointer to a uintptr and back, with arithmetic.
+>
+> ... Note that both [uintptr and unsafe.Pointer] conversions must appear in the
+> same expression, with only the intervening arithmetic between them:
+
+```go
+// INVALID: uintptr cannot be stored in variable
+// before conversion back to Pointer.
+u := uintptr(p)
+p = unsafe.Pointer(u + offset)
+```
+
+In a hypothetical future version of Go, it is possible that the value stored at
+the address pointed at by the `p` in the Go heap could be moved in between the
+conversion and assignment `u := uintptr(p)` and the pointer arithmetic `p =
+unsafe.Pointer(u + offset)`.
+
+Because we are relying on the invariant that the Go garbage collector (as of Go
+1.15) does not move data in the heap, [Brad
+Fitzpatrick](https://twitter.com/bradfitz) came up with a clever pseudo-package
+which users can underscore import if they intend to violate this particular
+package `unsafe` rule:
+[`go4.org/unsafe/assume-no-moving-gc`](https://pkg.go.dev/go4.org/unsafe/assume-no-moving-gc).
+If a program or its dependencies import this package and a future version of Go
+uses a moving garbage collector which may rearrange items in the heap, the
+package can be updated to make the program panic at runtime.
```

Comment:
I think it would be updated to outright fail to compile, rather than defer failure until runtime.